science/platform-implementations/scd40.cpp

```diff
@@ -0,0 +1,129 @@
+    
+    scd40_nm::scd40_settings get;
+     
+    get.temp_offset = 175 *(buffer[0] << 8 | buffer[1])/(1 << 16);
+
+    HAL_CHECK(hal::write(m_i2c,addresses::device_address, read_address_alt));
+    hal::delay(m_clock, 1ms);
+    HAL_CHECK(hal::read(m_i2c, addresses::device_address, buffer, hal::never_timeout()));
+    get.altitude = buffer[0]<<8 | buffer[1];
+
+    return get;
+}
+
+hal::status scd40_nm::set_settings( struct settings setting) {
+    if(setting.set_temp != 0){
+        int temp = int(((setting.set_temp*(1<<16)) / 175)) ;
+        hal::byte temp_first_half = (temp) >> 8;
+        hal::byte temp_second_half = temp & 0xff;
+        std::array<hal::byte, 4> set_temp_address = {set_temperature_offset_first_half, set_temperature_offset_second_half, temp_first_half, temp_second_half};
+        HAL_CHECK(hal::write(m_i2c,addresses::device_address,set_temp_address));
+        hal::delay(m_clock, 1ms);
+    }
+    if(setting.set_pressure != -1){
+        int pressure = int(setting.set_pressure/100);
+        hal::byte pressure_first_half = (pressure) >> 8;
+        hal::byte pressure_second_half = pressure & 0xff;
+        std::array<hal::byte, 4> set_pressure_address = {set_ambient_pressure_first_half, set_ambient_pressure_second_half, pressure_first_half, pressure_second_half};
+        HAL_CHECK(hal::write(m_i2c,addresses::device_address,set_pressure_address));
+        hal::delay(m_clock, 1ms);
+    } else if(setting.set_alt != 0){
+        int alt = int(setting.set_alt);
+        hal::byte alt_first_half = (alt) >> 8;
+        hal::byte alt_second_half = alt & 0xff;
+        std::array<hal::byte, 4> set_alt_address = {set_sensor_altitude_first_half, set_sensor_altitude_second_half, alt_first_half, alt_second_half};
+         HAL_CHECK(hal::write(m_i2c,addresses::device_address,set_alt_address));
+        hal::delay(m_clock, 1ms);
+    }
+
+    return hal::success();
+}
+
+hal::byte scd40_nm::generate_crc(std::array<hal::byte, 2> data){
+    uint16_t current_byte;
+    uint8_t crc_bit;
+    uint8_t crc = 0xFF;
+    uint8_t CRC8_POLYNOMIAL = 0x31;
+
+    for (current_byte = 0; current_byte < 2 ; ++current_byte) {
+        crc ^= (data[current_byte]);
+        for (crc_bit = 8; crc_bit > 0; --crc_bit) {
+            if (crc & 0x80)
+                crc = (crc << 1) ^ CRC8_POLYNOMIAL;
+            else
+                crc = (crc << 1);
+        }
+    }
+    return crc;
+}
+
+bool scd40_nm::validate_crc(std::array<hal::byte, 3> data) {
+    uint16_t current_byte;
+    uint8_t crc_bit;
+    uint8_t crc = 0xFF;
+    uint8_t CRC8_POLYNOMIAL = 0x31;
+
+    for (current_byte = 0; current_byte < 3 ; ++current_byte) {
+        crc ^= (data[current_byte]);
+        for (crc_bit = 8; crc_bit > 0; --crc_bit) {
+            if (crc & 0x80)
+                crc = (crc << 1) ^ CRC8_POLYNOMIAL;
+            else
+                crc = (crc << 1);
+        }
+    }
+    return crc == 0;
```

Comment:
This is fine for a first pass, but these should be pulled out into their own functions outside of this class. Consider the hypothetical situation where you have an application where 10 drivers require CRC generation. In this case, you end up paying for 10x implementations of the same CRC algorithm, resulting in a larger binary size for the application.

Reply:
Yes! @MichaelYKersey  mentioned this yesterday I wasn't sure if that was necessary since not all of our drivers do the CRC stuff. I suppose we can have a file in the implementations folder where we do CRC stuff.

Reply:
This is fine for now. You can resolve this comment. If y'all decide to add this to the libhal ecosystem, I'll add that the first step would be extracting the common stuff.